File: Java/47-PermutationsII.java
public class Solution {
    public List<List<Integer>> permuteUnique(int[] nums) {
        Arrays.sort(nums);
        List<List<Integer>> res = new ArrayList<>();
        List<Integer> curr = new ArrayList<>();
        if (nums.length == 0 || nums == null)   return res;
        boolean[] visited = new boolean[nums.length];
        permute(nums, res, curr, visited);
        return res;
    }
    private void permute(int[] nums, List<List<Integer>> res, List<Integer> curr, boolean[] visited){
        if (curr.size() == nums.length){
            res.add(new ArrayList<>(curr));
            return;
        }
        for (int i = 0; i < visited.length; i++){
            if (!visited[i]){
                if (i > 0 && nums[i] == nums[i-1] && visited[i-1]){
                    return;
                }
                visited[i] = true;
                curr.add(nums[i]);
                permute(nums, res, curr, visited);
                curr.remove(curr.size()-1);
                visited[i] = false;
            }
        }
    }
}

File: cpp/47-PermutationsII.cpp
//与I的区别在于有重复元素，所以在解集中要去重复。
//思路和combination II, subset II的去重复基本一致。通过排序
//每层递归跳过重复数字。注意这里的重复数字指的是一直到当前递归层，还未被使用的数字中的重复。
/*
https://discuss.leetcode.com/topic/46162/a-general-approach-to-backtracking-questions-in-java-subsets-permutations-combination-sum-palindrome-partioning
*/
class Solution {
public:
    vector<vector<int>> permuteUnique(vector<int>& nums) {
        vector<vector<int>> res;
        vector<int> curr;
        vector<bool> visited(nums.size(), false);
        if (nums.empty())
            return res;
        sort(nums.begin(), nums.end());
        permute(res, nums, curr, visited);
        return res;
    }
    void permute(vector<vector<int>>& res, vector<int>& nums, vector<int>& curr, vector<bool>& visited) {
        if (curr.size() == nums.size()) {
            res.push_back(curr);
            return;
        }
        for (int i = 0; i < nums.size(); i++) {
            if (visited[i]) {
                continue;
            }
            if (i != 0 && nums[i] == nums[i-1] && !visited[i-1]) {
                continue;
            }
            visited[i] = true;
            curr.push_back(nums[i]);
            permute(res, nums, curr, visited);
            visited[i] = false;
            curr.pop_back();
        }
    }
};
